src/topology/builder.rs

```diff
@@ -567,17 +558,13 @@ impl Runner {
             .input_rx
             .take()
             .expect("can't run runner twice")
-            .filter(move |event| ready(filter_event_type(event, self.input_type)))
+            .filter(move |events| ready(filter_events_type(events, self.input_type)))
             .ready_chunks(INLINE_BATCH_SIZE);
 
         self.timer.start_wait();
         while let Some(events) = input_rx.next().await {
             self.on_events_received(&events);
-
-            for event in events {
-                self.transform.transform(event, &mut outputs_buf);
-            }
-
+            self.transform.transform_all(events, &mut outputs_buf);
```

Comment:
Is it intended here that we're essentially multiplying the previous `ready_chunks` batch sizes by the new `EventArray` batch sizes? I'm not against increasing our buffer sizes by whatever amount shows good performance results, but it seems like it'd be a lot more clear to only have a single layer of batching than to continue using `ready_chunks` to queue up batches of batches. Then we could also likely leave this processing loop with the same structure as before and avoid introducing `transform_all`.

Reply:
No, it is not intended, but I'm not sure how best to resolve this at the moment. If you're ok with the rest of it, I'd like to defer that work to another issue where I'll outline the approaches I think are possible there.

Reply:
Sure, deferring seems reasonable. The simplest change I can think of would be to simply remove `ready_chunks` and let all batching happen via `EventArray`, but that doesn't really help us understand or control the size of those batches.

The questions I'd be curious to answer are roughly:

1. Are we using significantly more memory at peak with this change?
2. How much of the performance impact is due to simply buffering more events vs the actual change to `EventArray`?

Reply:
At this point, simply removing `ready_chunks` will result in moving just one event at a time again, at least until `SourceSender` and the sources have more work done to them.

The questions are good, and I don't have good answers for them. @blt do the soaks measure memory usage at any point? If the buffering makes this big a difference, does that indicate the `INLINE_BATCH_SIZE` constant is sized wrong?

Reply:
True, probably want to avoid introducing that potential regression for now. I guess the best path would be to first change `SourceSender` to use arrays, which should improve throughput, and then afterwards remove `ready_chunks`, which may reduce it somewhat? But then we'll be in a better position to evaluate buffer sizing without stacking potentially conflicting changes.

Reply:
I was also thinking of possibly writing a custom adapter that would be able to unify ready arrays up to a target limit size. This would help even out the flow into the subsequent stages.

Reply:
That's a good idea. You could maybe use [`ReadyFrames`](https://github.com/vectordotdev/vector/blob/0e140510090e6576d6ef6c6012f5084f2e8eff14/src/codecs/ready_frames.rs) as a starting point, since it has a roughly similar function.

Reply:
Thanks for the pointer @lukesteensen That was indeed a good starting point.